chips/sam4l/src/crccu.rs

```diff
@@ -0,0 +1,332 @@
+//! CRCCU implementation for the SAM4L
+//
+//  see datasheet section "41. Cyclic Redundancy Check Calculation Unit (CRCCU)"
+
+// Bugs:
+//
+// - Calculations with the two 32-bit polynomials produce results different
+//   from the reference in the notes below.
+
+// Infelicities:
+//
+// - As much as 512 bytes of RAM is wasted to allow runtime alignment of the
+//   CRCCU Descriptor.  Reliable knowledge of kernel alignment might allow this
+//   to be done statically.
+//
+// - CRC performance would be improved by using transfer-widths larger than Byte,
+//   but it is not clear in what cases that is possible.
+
+// TODO:
+//
+// - Chain computations to permit arbitrary-size computations, or at least
+//   publish the max buffer size the unit can handle.
+//
+// - Support continuous-mode CRC
+
+// Notes:
+//
+// http://www.at91.com/discussions/viewtopic.php/f,29/t,24859.html
+//      Atmel is using the low bit instead of the high bit so reversing
+//      the values before calculation did the trick. Here is a calculator
+//      that matches (click CCITT and check the 'reverse data bytes' to
+//      get the correct value):
+//
+//          http://www.zorc.breitbandkatze.de/crc.html
+//
+//      The SAM4L calculates 0x1541 for "ABCDEFG".
+
+use core::cell::Cell;
+use kernel::returncode::ReturnCode;
+use kernel::hil::crc::{self, Polynomial};
+use nvic;
+use pm::{Clock, HSBClock, PBBClock, enable_clock, disable_clock};
+
+// A memory-mapped register
+struct Reg(*mut u32);
+
+impl Reg {
+    fn read(self) -> u32 {
+        unsafe { ::core::ptr::read_volatile(self.0) }
+    }
+
+    fn write(self, n: u32) {
+        unsafe { ::core::ptr::write_volatile(self.0, n); }
+    }
+}
+
+// Base address of CRCCU registers.  See "7.1 Product Mapping"
+const CRCCU_BASE: u32 = 0x400A4000;
+
+// The following macro expands a list of expressions like this:
+//
+//    { 0x00, "Descriptor Base Register", DSCR, "RW" },
+//
+// into a series of items like this:
+//
+//    #[allow(dead_code)]
+//    const DSCR: Reg = Reg((CRCCU_BASE + 0x00) as *mut u32);
+
+macro_rules! registers {
```

Comment:
Hrm.... this is very very nice. On the other hand it violates the norm we've had of modeling the registers as a struct and putting a reference to it in the `Crc` struct (in this case).

If there is an easy way to re-purpose this macro to that model, that would really be ideal.